// ts-zone-file/lib/types.js
"use strict";
/**
 * Zone File Types
 */
Object.defineProperty(exports, "__esModule", { value: true });
var CAATAGENUM;
(function (CAATAGENUM) {
    CAATAGENUM[CAATAGENUM["issue"] = 0] = "issue";
    CAATAGENUM[CAATAGENUM["issuewild"] = 1] = "issuewild";
    CAATAGENUM[CAATAGENUM["iodef"] = 2] = "iodef";
})(CAATAGENUM = exports.CAATAGENUM || (exports.CAATAGENUM = {}));
var ZONETYPEENUM;
(function (ZONETYPEENUM) {
    ZONETYPEENUM["master"] = "master";
    ZONETYPEENUM["slave"] = "slave";
    ZONETYPEENUM["stub"] = "stub";
    ZONETYPEENUM["forward"] = "forward";
    ZONETYPEENUM["hint"] = "hint";
})(ZONETYPEENUM = exports.ZONETYPEENUM || (exports.ZONETYPEENUM = {}));
var TSIGALGORITHMENUM;
(function (TSIGALGORITHMENUM) {
    TSIGALGORITHMENUM["MD5"] = "hmac-md5";
    TSIGALGORITHMENUM["SHA1"] = "hmac-sha1";
    TSIGALGORITHMENUM["SHA224"] = "hmac-sha224";
    TSIGALGORITHMENUM["SHA256"] = "hmac-sha256";
    TSIGALGORITHMENUM["SHA384"] = "hmac-sha384";
    TSIGALGORITHMENUM["SHA512"] = "hmac-sha512";
})(TSIGALGORITHMENUM = exports.TSIGALGORITHMENUM || (exports.TSIGALGORITHMENUM = {}));
var AUTODNSSECENUM;
(function (AUTODNSSECENUM) {
    AUTODNSSECENUM["off"] = "off";
    AUTODNSSECENUM["allow"] = "allow";
    AUTODNSSECENUM["maintain"] = "maintain";
})(AUTODNSSECENUM = exports.AUTODNSSECENUM || (exports.AUTODNSSECENUM = {}));
